lean_source/07_Topology/source_01_Filters.lean

```diff
@@ -0,0 +1,465 @@
+specialized ``eventually_eq`` for the equality predicate. Note that both statement are
+definitionaly equivalent so the same proof will be used.
+
+TEXT. -/
+
+-- QUOTE:
+example (u v : ℕ → ℝ) (h : ∀ᶠ n in at_top, u n = v n) (x₀ : ℝ) : 
+  tendsto u at_top (𝓝 x₀) ↔ tendsto v at_top (𝓝 x₀) :=
+tendsto_congr' h
+
+example (u v : ℕ → ℝ) (h : u =ᶠ[at_top] v) (x₀ : ℝ) : 
+  tendsto u at_top (𝓝 x₀) ↔ tendsto v at_top (𝓝 x₀) :=
+tendsto_congr' h
+
+-- QUOTE.
+
+/- TEXT:
+
+It instructive to review the definition of filters in terms of this use with ``eventually``.
+Given ``F : filter X``, for any predicates ``P`` and ``Q`` on ``X``,
+
+* the condition ``univ ∈ F`` ensures that ``(∀ x, P x) → ∀ᶠ x in F, P x``.
+* the condition ``U ∈ F → U ⊆ V → V ∈ F`` ensures that ``(∀ᶠ x in F, P x) → (∀ x, P x → Q x) → ∀ᶠ x in F, Q x``
+* the condition ``U ∈ F → V ∈ F → U ∩ V ∈ F`` ensures that ``(∀ᶠ x in F, P x) → (∀ᶠ x in F, Q x) → ∀ᶠ x in F, P x ∧ Q x``, as we already discussed.
+
+TEXT. -/
+
+-- QUOTE:
+#check @eventually_of_forall
+#check @eventually.mono
+#check @eventually.and
+-- QUOTE.
+
+/- TEXT:
+The item above, corresponding to ``eventually.mono`` is also a very nice use of filters, especially combined
+with ``eventually.and``. There is also a tactic ``filter_upwards`` that allows to nicely combines these. 
+Compare:
+
+TEXT. -/
+
+-- QUOTE:
+
+example (P Q R : ℕ → Prop) (hP : ∀ᶠ n in at_top, P n) (hQ : ∀ᶠ n in at_top, Q n) 
+  (hR : ∀ᶠ n in at_top, P n ∧ Q n → R n) : 
+  ∀ᶠ n in at_top, R n := 
+begin
+  apply (hP.and (hQ.and hR)).mono,
+  rintros n ⟨h, h', h''⟩,
+  exact h'' ⟨h, h'⟩
+end
+
+example (P Q R : ℕ → Prop) (hP : ∀ᶠ n in at_top, P n) (hQ : ∀ᶠ n in at_top, Q n) 
+  (hR : ∀ᶠ n in at_top, P n ∧ Q n → R n) : 
+  ∀ᶠ n in at_top, R n := 
+begin
+  filter_upwards [hP, hQ, hR],
+  intros n h h' h'',
+  exact h'' ⟨h, h'⟩
+end
+
+-- QUOTE.
+
+/- TEXT:
+
+Readers who know about measure theory will note that the filter ``μ.ae`` of sets whose complement has measure zero
+(aka "the set of almost every point") is not very useful as the source or target of ``tendsto`` but very often
+used with ``eventually``, when saying a property holds for almost every point.
+
+There is dual version of ``∀ᶠ x in F, P x`` which is much less useful, but still exists: ``∃ᶠ x in F, P x`` means
+``{x | ¬P x} ∉ f``. For instance ``∃ᶠ n in at_top, P n`` means there are arbitrarily large ``n`` such that ``P n``.
+This notation stands for ``filter.frequently``.
+
+
+As a more sophisticated example, the statement: "if a sequence ``u`` converges to
+some ``x`` and ``u n`` belongs to a set ``M`` for ``n`` large enough then ``x`` is in the closure of
+``M``" is formalized as: ``tendsto u at_top (𝓝 x) → (∀ᶠ n in at_top, u n ∈ M) → x ∈ closure M``,
+which is a special case of ``mem_closure_of_tendsto`` from the topology library.
+See if you can prove it using the quoted lemmas, knowing that ``cluster_pt x F`` means ``(𝓝 x ⊓ F).ne_bot``.
```

Comment:
```suggestion
See if you can prove it using the quoted lemmas, keeping in mind that ``cluster_pt x F`` means ``(𝓝 x ⊓ F).ne_bot``.
```

Reply:
It doesn't seem fair to write "keeping in mind" when this has never been explained.